lib: os: timespec_util: reimplement and test with broader inputs

cfriedt opened the pull request:
Originally, the `timespec_to_timeout()` and `timespec_from_timeout()` tests assumed that `tv_sec` and `tv_nsec` values divided evenly with the number of ticks per second (`CONFIG_SYS_CLOCK_TICKS_PER_SEC`). However, those assumptions broke down when testing with 32768 ticks/s .

Upon further investigation, there were additional corner cases discovered that were not handled in an ideal or safe way.

Part of this fix involved identifying several "domains" and special values that have semantic meaning that must be handled particularly carefully.

Fixes #92158
Fixes #92908
Fixes #92912
Fixes #90479 (Coverity CID 524770)

Some context for #92908 and #92912 is provided below.

## Added Constants

Additional hidden-API constants were made to simplify the solution. These are "hidden" via an existing Doxygen `@cond` block.

`K_TICK_MIN`: 1
`K_TICK_MAX`: `UINT32_MAX`-1 in 32-bit, `INT64_MAX` in 64-bit

`K_TIMESPEC_MIN`: the equivalent of `K_TICK_MIN` as a `timespec` object
`K_TIMESPEC_MAX`: the equivalent of `K_TICK_MAX` as a `timespec` object

`K_TIMESPEC_NO_WAIT`: the equivalent of `K_NO_WAIT` as a `timespec` object
`K_TIMESPEC_FOREVER`: the equivalent of `K_FOREVER` as a `timespec` object

## Coverage Details

1. Converting a negative `k_timeout_t`

These only exist in 64-bit representation and actually encode absolute timepoints via ticks. Since the stated purpose of the conversion functions is to convert between durations, we must reject absolute time points in timespec_from_timeout(). The previous implementation did this as well.

2. Converting a negative `timespec`

We assume that this duration means a timeout has already expired, and round these up to `K_NO_WAIT` in `timespec_to_timeout()`. The previous implementation also considered this case but it has been simplified with `timespec_compare()`.

3. Due to the larger numeric space in the timespec representation, the reverse mapping of timespec to `k_timeout_t` is "fuzzy". However, `K_NO_WAIT` (`K_TIMESPEC_NO_WAIT`) and `K_FOREVER` (`K_TIMESPEC_FOREVER`) must remain semantically equivalent. The previous implementation also held this to be true, but the test cases are a bit clearer about it now.

4. Also, due to the larger numeric space in timespec representation, there was a special requirement to round up to the nearest tick boundary for any timespec in the strictly exclusive range (`K_TIMESPEC_NO_WAIT`, `K_TIMESPEC_MAX`). We must round up, since
  * rounding down to `K_NO_WAIT` is most certainly not representative of a non-zero finite duration delay
  * the kernel operates on tick boundaries

However, since it's possible that rounding up might not _always_ be the right thing to do, in order to allow the application to make more informed decisions, we created a modified
`timespec_to_timeout_rem()` that also returns the remainder (or difference) between the requested time to convert and resulting `k_timeout_t`. The difference is expressed as a `timespec` object. The previous implementation did not provide this capability in a convenient manner.

5. Above the `K_TIMESPEC_MAX` boundary, which is the highest possible timespec that can be represented by a tick, and specifically in the 64-bit timeout representation, there is a domain that cannot be rounded up to `K_TIMESPEC_FOREVER` and should always be rounded down to `K_TIMESPEC_MAX`.

This is to ensure that finite durations remain finite (even if they are beyond the heat death of the universe).

## Added API

```cpp
k_timeout_t timespec_to_timeout_rem(const struct timespec *req, struct timespec *rem)
```

The function above accounts for the fact that converting from `timespec` (which has a much larger numerical range) to `k_timeout_t` is a "fuzzy" conversion. The entire numerical range of `timespec` must be mapped to either a semantic `k_timeout_t` or must be aligned to an appropriate tick boundary.

Fortunately, it is not a "lossy" conversion.

## Inline Functions changed

* `timespec_from_timeout()` was changed from an inline to a regular function
* `timespec_to_timeout()` was changed to a wrapper around `timespec_to_timeout_rem()` with `rem = NULL`

## Testing Done

Additional test cases were defined to exercise permutations of:
* 32-bit platform
* 64-bit platform
* CONFIG_SPEED_OPTIMIZATIONS = y / n
* CONFIG_TIMEOUT_64BIT = y / n
* low tick rate (1 Hz)
* high tick rate (1000000 Hz)
* 32k tick rate (32768 Hz)
* Prime tick rate (10007 Hz)

```shell
$ twister -T tests/lib/timespec_util/
INFO    - Using Ninja..
INFO    - Zephyr version: v4.2.0-rc2-36-g05ebc9fc942f
INFO    - Using 'zephyr' toolchain.
INFO    - Selecting default platforms per testsuite scenario
INFO    - Building initial testsuite list...
INFO    - Writing JSON report /home/cfriedt/zephyrproject/zephyr/twister-out/testplan.json
INFO    - JOBS: 32
INFO    - Adding tasks to the queue...
INFO    - Added initial list of jobs to queue
INFO    - Total complete:   18/  18  100%  built (not run):    0, filtered:   86, failed:    0, error:    0
INFO    - 12 test scenarios (104 configurations) selected, 86 configurations filtered (86 by static filter, 0 at runtime).
INFO    - 18 of 18 executed test configurations passed (100.00%), 0 built (not run), 0 failed, 0 errored, with no warnings in 26.27 seconds.
INFO    - 198 of 198 executed test cases passed (100.00%) on 9 out of total 1115 platforms (0.81%).
INFO    - 18 test configurations executed on platforms, 0 test configurations were only built.
INFO    - Saving reports...
INFO    - Writing JSON report /home/cfriedt/zephyrproject/zephyr/twister-out/twister.json
INFO    - Writing xunit report /home/cfriedt/zephyrproject/zephyr/twister-out/twister.xml...
INFO    - Writing xunit report /home/cfriedt/zephyrproject/zephyr/twister-out/twister_report.xml...
INFO    - Run completed
```

cfriedt commented on include/zephyr/sys/clock.h:
Note: the declarations added in this section are within hidden blocks

```cpp
/** @cond INTERNAL_HIDDEN */
...
/** @endcond */
```

cfriedt commented:
If all tests are passing, I would strongly recommend merging this before `v4.2.0` is tagged. If that is not posible, I would strongly recommend backporting to `v4.2-branch` as soon as possible.

aescolar reviewed (changes requested):
If I understand the code correctly there is some overflow issues.
Note also a few suggestions so we would be able to support other time_t sizes.
(I have not reviewed the test itself.)

cfriedt commented:
TODO:
- [ ] change the upper bound for high tick rate to 1000000
- [ ] rename test configurations to be clearer
- [ ] define `SYS_TIME_T_MIN` / `SYS_TIME_T_MAX`

aescolar commented on lib/os/timeutil.c:
```diff
@@ -0,0 +1,85 @@
+		*ts = (struct timespec){
+			.tv_sec =
+				(time_t)((uint64_t)timeout.ticks / CONFIG_SYS_CLOCK_TICKS_PER_SEC),
+			.tv_nsec = (long)(K_TICKS_TO_NSECS((uint64_t)timeout.ticks) % NSEC_PER_SEC),
```
> .. all of the nanosecond information is encoded in the least-significant bits.

It is not.
1024 mod 100 = 24
2048 mod 100 = 48
...
Every bit is meaningful in a division unless the divisor is a power of 2.
Only if the divisor was a power of 2, would that be correct, and in that case you would only need to mask out the lower bits instead of doing a module operation.

aescolar replied:
> ??

@cfriedt I tried to provide you a simple example with few digits which demonstrates that your understanding is incorrect.
That is the example being that numbers in binary with only higher bits set and changing (in this case bits 10 and 11), when divided by decimal 100 (or in that case the remainder of the divison) do not yield the same result just because the divisor can be represented with fewer bits than the position of the changing bits in the dividend.

I hoped you would have been able to extrapolate the logic of that demonstration to bits above the 63th and a divisor of 1e9 instead of 100.

That is, I hoped that you would have understood that your statement "all of the nanosecond information is encoded in the least-significant bits" is incorrect.

> since all of the nanosecond information is encoded in the least-significant bits. ... I apply the modulus of 1e9 at the end (i.e. discard information in the most significant bits at the end of the computation).

That is not how representing numbers works.. Neither how division works..

I will try to be a bit more verbose, though I had hoped to save the time.
A number is represented in a base.
Each digit of a number is a part of that number.
Where, in base 'b', counting digits from the right, digit n corresponds to a part of that number, corresponding to multiplying that digit by (b^n) (where '^' is exponent, 'n' starts at 0)
https://en.wikipedia.org/wiki/Numeral_system#Positional_systems_in_detail

In binary, the base b is 2, and we call each digit a bit.

Note how 4096(base10) in base 2 is 0b1000000000000, and how with one digit more 0b10000000000000 it is 8192 (base10).
Note that even though only relatively high bits are changing, in base 10 the lower digits are changing.

Similarly if we now look at the next to the maximum number representable in int64_t (I will use hex notation for brevity)
0x8000000000000000 == 9223372036854775808 (base10)
and one bit higher 0x10000000000000000 = 18446744073709551616 (base10).

Note how those higher bits carry information about something changing below 1e9.

Now let's take the current definition of K_TS_MAX, and its nanosecond part in particular.
K_TICK_MAX is INT64_MAX (0x7FFFFFFFFFFFFFFF)
In your calculation:
((ticks * NSEC_PER_SEC) / CONFIG_SYS_CLOCK_TICKS_PER_SEC) % NSEC_PER_SEC),
Let's just assume (CONFIG_SYS_CLOCK_TICKS_PER_SEC = 100 for this example)

With arbitrary bitwidth that is:
0x7FFFFFFFFFFFFFFF * 1e9(dec) = 0x1DCD64FFFFFFFFFFC4653600 (9223372036854775807000000000 dec)
0x1DCD64FFFFFFFFFFC4653600 / 100(dec)= 0x4C4B3FFFFFFFFFFF676980
0x4C4B3FFFFFFFFFFF676980 % 1e9 = 0x42C1D80 (70000000 dec).

If instead we limit ourselves to 64bits dropping the most significant bits:
0x7FFFFFFFFFFFFFFF * 1e9(dec) masked = 0xFFFFFFFFC4653600 (18446744072709551616 dec)
0xFFFFFFFFC4653600 / 100 (dec) (integer division) = 0x28F5C28F529F8DC (184467440727095516 decimal)
0x28F5C28F529F8DC % 1e9 = 0x2B5698DC (727095516 decimal)

Note how the operation yielded a different result due to discarding the MSBs.
Note that even though our clock tick period is 10ms (10000000 ns), we ended up with a convertion from ticks to time that is not a multiple of the tick period.

> and you are missing parentheses.

I don't think so. Please note that C follows the same conventions for '*' '/' operator precendence as mathematics, that is left to right.
https://en.cppreference.com/w/c/language/operator_precedence.html

> If you're confident that my approach is incorrect in spite of the full spectrum of numerical inputs that prove otherwise, then please feel free to suggest a testcase that fails.

Here a simple "testcase", where we simply print out the nanosecond part of the constant K_TS_MAX and see it is not possibly correct.
This example code then shows how the code I proposed does instead produce the correct result:

```diff
diff --git a/samples/hello_world/src/main.c b/samples/hello_world/src/main.c
index c550ab461cb..2a5c33575db 100644
--- a/samples/hello_world/src/main.c
+++ b/samples/hello_world/src/main.c
@@ -5,10 +5,36 @@
  */
 
 #include <stdio.h>
+#include <time.h>
+#include <zephyr/sys/clock.h>
+
+void correct_calc(k_timeout_t timeout, struct timespec *ts) {
+       uint64_t secs = (uint64_t)timeout.ticks / CONFIG_SYS_CLOCK_TICKS_PER_SEC;
+       uint64_t frac_ticks = (uint64_t)timeout.ticks - secs*CONFIG_SYS_CLOCK_TICKS_PER_SEC;
+
+       *ts = (struct timespec){
+               .tv_sec = secs,
+               .tv_nsec = frac_ticks*1000000000ULL/CONFIG_SYS_CLOCK_TICKS_PER_SEC
+       };
+}
 
 int main(void)
 {
        printf("Hello World! %s\n", CONFIG_BOARD_TARGET);
 
+       struct timespec t = K_TS_MAX;
+
+       printf("nsec = %li\n", t.tv_nsec);
+
+       struct timespec t2;
+       k_timeout_t tick2;
+       tick2.ticks = K_TICK_MAX;
+
+       correct_calc(tick2, &t2);
+
+       printf("Correct nsec = %li\n", t2.tv_nsec);
+       printf("CONFIG_SYS_CLOCK_TICKS_PER_SEC = %u\n", CONFIG_SYS_CLOCK_TICKS_PER_SEC);
+       printf("Sys tick period in ns = %lli\n", 1000000000ULL/CONFIG_SYS_CLOCK_TICKS_PER_SEC);
+
        return 0;
 }
```

which produces for native_sim//64:

```
*** Booting Zephyr OS build v4.2.0-rc1-225-g9ccd24fdfec6 ***
Hello World! native_sim/native/64
nsec = 727095516
Correct nsec = 70000000
CONFIG_SYS_CLOCK_TICKS_PER_SEC = 100
Sys tick period in ns = 10000000
```

cfriedt replied:
@aescolar - thank you for clarifying.

I think the miscommunication in this case is that nanosecond information is only contained within the least-significant bits when we're dealing with ticks. Once the time representation changes, it is spread through all of the bits.

I see now why you were so adamantly disagreeing about this.

> 1024 mod 100 = 24
> 2048 mod 100 = 48

The counterexample you presented originally did not seem to support the point you were trying to make whatsoever.

Thank you for providing a concrete example to illustrate your point - which is really arithmetic in nature.

cfriedt commented on lib/os/timeutil.c:
```diff
@@ -0,0 +1,85 @@
+	__ASSERT_NO_MSG(timespec_is_valid(ts));
+}
+
+static k_timeout_t timespec_timeout_rem(const struct timespec *ts, struct timespec *rem,
```
Fixed!

The doxygen for `timespec_from_timeout()` already includes the pseudo-arithmetic expression.

cfriedt commented:
* removed unnecessary comment

cfriedt commented on lib/os/timeutil.c:
```diff
@@ -0,0 +1,85 @@
+/*
```
It's not deleted sorry - just added it's contents to `lib/utils/timeutil.c` I'll remove it now.

cfriedt commented on include/zephyr/sys/timeutil.h:
```diff
@@ -517,10 +497,8 @@ static inline bool timespec_negate(struct timespec *ts)
-	    ((sizeof(ts->tv_sec) == sizeof(int64_t)) && (ts->tv_sec == INT64_MIN))) {
-		/* -INT64_MIN > INT64_MAX, so +ve integer overflow would occur */
+	if (ts->tv_sec == SYS_TIME_T_MIN) {
+		/* -SYS_TIME_T_MIN > SYS_TIME_T_MAX, so +ve integer overflow would occur */
```
I assume this is fixed?

Commit:
Revert "tests/lib/timespec_util: Skip two tests by now"

This reverts commit 8f092623aee24c5f8ed81798f81a05759a081bfa.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

andyross commented on include/zephyr/sys/clock.h:
```diff
@@ -166,6 +166,57 @@ typedef struct {
+	(((uint64_t)(ticks) >= (uint64_t)K_TICKS_FOREVER)                                          \
+		 ? (NSEC_PER_SEC - 1)                                                              \
+		 : ((((uint64_t)(ticks) % CONFIG_SYS_CLOCK_TICKS_PER_SEC) * NSEC_PER_SEC) /        \
+		    CONFIG_SYS_CLOCK_TICKS_PER_SEC))
```
Any reason why you don't want to implement these in terms of k_ticks_to_ns_xxx() from time_units.h?  Those have been pretty battle-hardened over the years, there are lots of edge cases with these conversions.  And they should (at least they used to) expand cleanly in a constant context to eliminate a runtime call when needed.

cfriedt replied:
Yeah, I can do that and tests still pass. Please let me know if you see anything else that should be changed.

```diff
diff --git a/include/zephyr/sys/clock.h b/include/zephyr/sys/clock.h
index 9744201d99e..29dadddb07b 100644
--- a/include/zephyr/sys/clock.h
+++ b/include/zephyr/sys/clock.h
@@ -172,16 +172,14 @@ typedef struct {
 
 /* Converts ticks to seconds */
 #define K_TICKS_TO_SECS(ticks)                                                                     \
-       (((uint64_t)(ticks) >= (uint64_t)K_TICKS_FOREVER)                                          \
-                ? SYS_TIME_T_MAX                                                                  \
-                : ((uint64_t)(ticks) / CONFIG_SYS_CLOCK_TICKS_PER_SEC))
+       (((uint64_t)(ticks) >= (uint64_t)K_TICKS_FOREVER) ? SYS_TIME_T_MAX                         \
+                                                         : k_ticks_to_sec_floor64(ticks))
 
 /* Converts ticks to nanoseconds, modulo NSEC_PER_SEC */
 #define K_TICKS_TO_NSECS(ticks)                                                                    \
        (((uint64_t)(ticks) >= (uint64_t)K_TICKS_FOREVER)                                          \
                 ? (NSEC_PER_SEC - 1)                                                              \
-                : ((((uint64_t)(ticks) % CONFIG_SYS_CLOCK_TICKS_PER_SEC) * NSEC_PER_SEC) /        \
-                   CONFIG_SYS_CLOCK_TICKS_PER_SEC))
+                : k_ticks_to_ns_floor32((uint64_t)(ticks) % CONFIG_SYS_CLOCK_TICKS_PER_SEC))
 
 /* Define a timespec */
 #define K_TIMESPEC(sec, nsec)
```

cfriedt commented on include/zephyr/sys/clock.h:
```diff
@@ -166,6 +166,44 @@ typedef struct {
+/* The semantic equivalent of K_FOREVER but expressed as a timespec object*/
+#define K_TS_FOREVER                                                                               \
+	((struct timespec){                                                                        \
+		.tv_sec = (time_t)INT64_MAX,                                                       \
```
I think I've made all of those substitutions at this point.

I ended up removing line 389, but the `uint32_t` was correct (since a `long` would be 8 bytes otherwise on 64-bit machines).

aescolar replied:
> I think I've made all of those substitutions at this point.

Thank you.

> (since a long would be 8 bytes otherwise on 64-bit machines).

Note that is not true in general. It depends on the architecture and compiler. See for ex. https://en.wikipedia.org/wiki/64-bit_computing#64-bit_data_models

Commit:
sys: clock: additional constants and macros for timespec conversion

Define additional constants and macros for converting between timespec
objects and k_timeout_t.

There are four semantically important durations identified with new
constants identified via '.

1. K_NO_WAIT / K_TIMESPEC_NO_WAIT'
  *  duration of zero

2. K_TICK_MIN' / K_TIMESPEC_MIN'
  * minimum number of ticks strictly greater than K_NO_WAIT

3. K_TICK_MAX' / K_TIMESPEC_MAX'
  * maximum number of ticks strictly and semantically less than K_FOREVER

4. K_FOREVER / K_TIMESPEC_FOREVER'
  * the familiar non-expiring number of ticks (i.e. infinity)

With these definitions, there are 3 meaningful regions for conversion
between timespec and k_timeout_t representation:

1. < K_TIMESPEC_NO_WAIT (i.e. a negative timespec)
  * should be rounded up to K_TIMESPEC_NO_WAIT
  * < K_NO_WAIT in ticks is an absolute time point in 64-bit
    representation should no be permitted in the affected conversion
    routines, as they are explicitly  for converting time durations rather
    than absolute time points.

2. K_TIMESPEC_NO_WAIT < x < K_TIMESPEC_MAX
  * should be rounded-up to the next tick boundary
  * must be represented as the minimal non-zero duration in ticks

3. K_TIMESPEC_MAX <= x < K_TIMESPEC_FOREVER
  * must be represented as the maximum expiring duration in ticks
  * should be rounded-down to K_TIMESPEC_MAX

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
tests: lib: timespec_util: elaborate timespec to k_timeout_t conversion

Elaborate on testing for functions that convert between timespec and
k_timeout_t.

We explicitly add tests to verify functionality in the regions of
interest, verify the semantic equivalence of specific time points.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
sys: timeutil: ensure safety of timespec conversion routines

Originally, the timespec_to_timeout() and timespec_from_timeout() tests
assumed that tv_sec and tv_nsec values divided evenly with the number of
ticks per second (CONFIG_SYS_CLOCK_TICKS_PER_SEC). However, those
assumptions broke down when testing with 32768 ticks/s .

Upon further investigation, there were additional corner cases
discovered that were not handled in an ideal or safe way.

Part of this fix involved identifying several "domains" and special
values that have semantic meaning that must be handled particularly
carefully.

Additional hidden-API constants were made to simplify the solution.

K_TICK_MIN (1)
K_TICK_MAX (UINT32_MAX-1 in 32-bit, INT64_MAX in 64-bit)

K_TIMESPEC_NO_WAIT (like K_NO_WAIT)
K_TIMESPEC_FOREVER (like K_FOREVER)

1. Converting a negative k_timeout_t

These only exist in 64-bit representation and actually encode absolute
timepoints via ticks. Since the stated purpose of the conversion
functions is to convert between durations, we must reject absolute
time points in timespec_from_timeout().

2. Converting a negative timespec

We assume that this duration means a timeout has already expired, and
round these up to K_NO_WAIT in timespec_to_timeout().

3. Due to the larger numeric space in the timespec representation,
the reverse mapping of timespec to k_timeout_t is "fuzzy". However,
K_NO_WAIT (K_TIMESPEC_NO_WAIT) and K_FOREVER (K_TIMESPEC_FOREVER) must
remain semantically equivalent. The previous implementation also held this
to be true, but the test cases are a bit clearer about it now.

4. Also, due to the larger numeric space in timespec representation,
there was a special requirement to round up to the nearest tick
boundary for any timespec in the strictly exclusive range
(K_TIMESPEC_NO_WAIT, K_TIMESPEC_MAX). We must round up, since
a) rounding down to K_NO_WAIT is most certainly not representative
   of a non-zero finite duration delay
b) the kernel operates on tick boundaries

5. Above the K_TIMESPEC_MAX boundary, which is the highest possible
timespec that can be represented by a tick, and specifically in the
64-bit timeout representation, there is a domain that cannot be
rounded up to K_TIMESPEC_FOREVER and should always be rounded down to
K_TIMESPEC_MAX.

This is to ensure that finite durations remain finite (even if
they are beyond the heat death of the universe).

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
sys: timeutil: compute the remainder in timespec to timeout conversion

Since it's possible that rounding up might not always be the right thing
to do in every situation, in order to allow the application to make more
informed decisions, we created a modified timespec_to_timeout() that also
returns the remainder (or difference) between the requested time to
convert and resulting k_timeout_t. The difference is expressed as a
timespec object.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

aescolar commented on tests/lib/timespec_util/src/main.c:
```diff
@@ -329,54 +399,135 @@ ZTEST(timeutil_api, test_timespec_from_timeout)
+#if defined(CONFIG_TIMEOUT_64BIT) && (CONFIG_SYS_CLOCK_TICKS_PER_SEC == 100)
+	{
+		struct timespec rem = {};
+		k_timeout_t to = K_TICKS(9223372036854775807LL);
```
```suggestion
		k_timeout_t to = K_TICKS(K_TICK_MAX);
```
This is just `K_TICKS_MAX`, why not use its name?

cfriedt replied:
I'd be happy to add a comment.

cfriedt replied:
Fixed!

cfriedt replied:
I'll replace with the constant. No sense in dying on that hill.

aescolar commented on tests/lib/timespec_util/src/main.c:
```diff
@@ -329,54 +399,135 @@ ZTEST(timeutil_api, test_timespec_from_timeout)
+			     tspec->tspec.tv_nsec);
+	}
+
+	/* Special case highlighed in #92709 */
```
```suggestion
	/* Check that K_TICK_MAX is properly converted and matches into K_TS_MAX */
```
This is not a special case, just a check of the conversion of `K_TICK_MAX` 

Could you also check that and `K_TS_MAX` matches?

cfriedt replied:
Fixed!

aescolar commented on tests/lib/timespec_util/src/main.c:
```diff
@@ -329,54 +425,138 @@ ZTEST(timeutil_api, test_timespec_from_timeout)
+	{
+		struct timespec rem = {};
+		k_timeout_t to = K_TICKS(9223372036854775807LL);
+		struct timespec ts = K_TIMESPEC(92233720368547758L, 70000000L);
```
```suggestion
		struct timespec ts = K_TIMESPEC(92233720368547758LL, 70000000L);
```
That number needs 40bits.

aescolar reviewed (commented):
I can see you are pushing updates right now, so I will wait for you.

aescolar commented on tests/lib/timespec_util/src/main.c:
```diff
@@ -329,54 +423,138 @@ ZTEST(timeutil_api, test_timespec_from_timeout)
+	PRINT_LINE;
+
+	zassert_equal(K_TICK_MAX,
+		      IS_ENABLED(CONFIG_TIMEOUT_64BIT) ? 9223372036854775807LL : 4294967294);
```
```suggestion
		      IS_ENABLED(CONFIG_TIMEOUT_64BIT) ? INT64_MAX : UINT32_MAX);
```

cfriedt replied:
That is incorrect
```cpp
IS_ENABLED(CONFIG_TIMEOUT_64BIT) ? INT64_MAX : UINT32_MAX);
```

This is one of the reasons I would kind of prefer to keep the numerical values there.

I would be happy to add a comment though.

aescolar replied:
> The above is checked symbolically in quite a few places already. I wanted to include a concrete numeric value too almost specifically for this purpose.

We can trust this time_util test that INT64_MAX and UINT32_MAX are what it needs to be..

cfriedt replied:
Sure

aescolar commented on tests/lib/timespec_util/src/main.c:
```diff
@@ -329,54 +423,138 @@ ZTEST(timeutil_api, test_timespec_from_timeout)
+		zexpect_true(timespec_equal(&rem, &K_TS_NO_WAIT), "non-zero remainder {%lld, %lld}",
+			     (long long)rem.tv_sec, (long long)rem.tv_nsec);
+
+		TC_PRINT("timespec_to_timeout_rem():\nts: {%lld, %lld} => to: {%" PRId64
```
```suggestion
		TC_PRINT("timespec_to_timeout_rem():\nts: {%lld, %lld} => to: {%lld
```

aescolar commented on tests/lib/timespec_util/src/main.c:
```diff
@@ -268,54 +264,152 @@ ZTEST(timeutil_api, test_timespec_equal)
+	DECL_NSAT_TOSPEC_TEST(K_TIMESPEC(0, K_TICKS_TO_NSECS(K_TICK_MIN))),
+
+#if CONFIG_SYS_CLOCK_TICKS_PER_SEC < GHZ(1)
+	DECL_NSAT_TOSPEC_TEST(K_TIMESPEC(0, 1)),
```
This line is a rep of 348? so I guess instead 348 and 349 should be here?

cfriedt replied:
I think that the GHZ condition (and likely the macro definition) can probably be removed. It's highly unlikely that we'll ever have a tick rate that's >= 1 GHz

cfriedt replied:
Fixed!

aescolar commented on tests/lib/timespec_util/src/main.c:
```diff
@@ -329,54 +423,138 @@ ZTEST(timeutil_api, test_timespec_from_timeout)
+				      i, (int64_t)actual.ticks, (int64_t)K_TICK_MIN);
+		} else if (tspec->saturation > 0) {
+			/*
+			 * Careful - the compiler can play tricks!
```
what is all this?

cfriedt replied:
I wish I knew!

aescolar replied:
After fixing a few obvious bugs in the test code https://github.com/aescolar/zephyr/commit/fc8e3fb44b2c14ce5267d2e131036f25116f93ee this part seems to work fine without this weird workaround.

cfriedt commented on tests/lib/timespec_util/src/main.c:
```diff
@@ -100,16 +95,15 @@ static const struct ts_test_spec ts_tests[] = {
-	DECL_INVALID_TS_TEST(INT64_MIN + 2, -2 * NSEC_PER_SEC - 1, 0, 0, UNCORRECTABLE),
-	DECL_INVALID_TS_TEST(INT64_MIN + 1, -NSEC_PER_SEC - 1, 0, 0, UNCORRECTABLE),
-	DECL_INVALID_TS_TEST(INT64_MIN + 1, -NSEC_PER_SEC - 1, 0, 0, UNCORRECTABLE),
+	DECL_INVALID_TS_TEST(INT64_MIN + 2, -2 * (int64_t)NSEC_PER_SEC - 1, 0, 0, UNCORRECTABLE),
```
> In all these should INT64_MIN be replaced with SYS_TIME_T_MIN and INT64_MAX w SYS_TIME_T_MAX?

They should not.

This testsuite requires a 64-bit `time_t`, which corresponds to AFAIK, every actual platform that Zephyr supports aside from `native_sim` when built with the host C library forcing 32-bit mode. That corner case is filtered out in YAML.

> why the cast to int64_t? (also in the next 5 lines)

because otherwise, the tests would fail (previously `NSEC_PER_SEC` was undefined and redefined as a signed type to account for these cases). The local cast for these test cases is a bit less invasive.

cfriedt replied:
Regressing in this case just means moving backward.

The testsuite also has never supported 32-bit `time_t` so doing so is a new feature and really has nothing to do with the bugs that are being fixed with this PR.

Additionally, It's not that it couldn't be done. It's that it really shouldn't. We should be moving in the direction of universally adopting a 64-bit `time_t`.

> * MISRA guidelines, specifically MISRA-C:2012 Rule 21.10 and MISRA-CPP:2008 Rule 18.0.4, discourage the use of functions from time.h.
> * This is because these functions have unspecified or implementation-dependent behavior, meaning their behavior can vary depending on the compiler and target platform.
> * When a 32-bit time_t is used, it introduces the year 2038 problem, making the use of these functions problematic. 

> It is just writing portable and understandable code.

What is portable about `SYS_TIME_T_MIN` and `SYS_TIME_T_MAX`?

> Still applies

Does it?

cfriedt replied:
The question about this code change has already been answered in a previous comment, so I'm marking it as resolved.

```cpp
DECL_INVALID_TS_TEST(INT64_MIN + 2, -2 * (int64_t)NSEC_PER_SEC - 1, 0, 0, UNCORRECTABLE),
```

> > why the cast to int64_t? (also in the next 5 lines)

> because otherwise, the tests would fail (previously NSEC_PER_SEC was undefined and redefined as a signed type to account for these cases). The local cast for these test cases is a bit less invasive.

aescolar replied:
> In all these should INT64_MIN be replaced with SYS_TIME_T_MIN and INT64_MAX w SYS_TIME_T_MAX?
(from lines 77 to 106)
(same comment for test_timespec_add 163-166)

This is the request for changes which was not done.

cfriedt commented on tests/lib/timespec_util/src/main.c:
```diff
@@ -329,54 +423,138 @@ ZTEST(timeutil_api, test_timespec_from_timeout)
+	}
+
+	/* Special case highlighed in #92709 */
+#if defined(CONFIG_TIMEOUT_64BIT) && (CONFIG_SYS_CLOCK_TICKS_PER_SEC == 100)
```
It shouldn't need to be skipped.

`time_t` is enforced to be 64-bits in this testsuite. Given a 64-bit k_timeout_t and a tick rate of 100Hz that check should be valid every single time.

cfriedt replied:
See https://github.com/zephyrproject-rtos/zephyr/pull/92709/files#r2199259816

cfriedt replied:
I think, given that it has been demonstrated that this test does not need to be skipped, I'm going to mark this conversation as resolved.

cfriedt commented on tests/lib/timespec_util/src/main.c:
```diff
@@ -100,16 +95,15 @@ static const struct ts_test_spec ts_tests[] = {
 	DECL_INVALID_TS_TEST(0, LONG_MIN, LONG_MAX / NSEC_PER_SEC, 145224192, CORRECTABLE),
-	DECL_INVALID_TS_TEST(0, LONG_MAX, LONG_MAX / NSEC_PER_SEC, LONG_MAX % NSEC_PER_SEC,
-			     CORRECTABLE),
+	DECL_INVALID_TS_TEST(0, LONG_MAX, LONG_MAX / NSEC_PER_SEC, LONG_MAX, CORRECTABLE),
```
`timespec_normalize()` will actually assert if the timespec is invalid, but yeah, for invalid tests here, there is no check made against "valid_ts", as mentioned in the comments above.

cfriedt replied:
> I do not know which comments you refer to

The comments above this code, where `struct ts_test_spec` is declared.

```cpp
/*
 * test spec for simple timespec validation
 *
 * If a timespec is expected to be valid, then invalid_ts and valid_ts are equal.
 *
 * If a timespec is expected to be invalid, then invalid_ts and valid_ts are not equal.
 */
```

The values are expected to simply be unequal.

```cpp
DECL_INVALID_TS_TEST(0, LONG_MAX, LONG_MAX / NSEC_PER_SEC, LONG_MAX, CORRECTABLE),
```

The test above simply asserts that the un-normalized timespec is correctable (the assertion is done in `timespec_normalize()`)

> shouldn't the result of the normalization be checked against valid_ts?

Possibly in a future PR. It's unrelated to this change.

cfriedt commented on include/zephyr/sys/clock.h:
```diff
@@ -166,6 +166,55 @@ typedef struct {
+#define SYS_TIME_T_MAX ((((time_t)1 << (8 * sizeof(time_t) - 2)) - 1) * 2 + 1)
+#define SYS_TIME_T_MIN (-SYS_TIME_T_MAX - 1)
+
+/* Converts ticks to seconds */
```
Fixed!

aescolar commented on include/zephyr/sys/clock.h:
```diff
@@ -166,6 +166,55 @@ typedef struct {
+		 : k_ticks_to_ns_floor32((uint64_t)(ticks) % CONFIG_SYS_CLOCK_TICKS_PER_SEC))
+
+/* Define a timespec */
+#define K_TIMESPEC(sec, nsec)                                                                      \
```
I would think it is best to move it, but the @andyross  or @nashif can decide

cfriedt replied:
Andy previously approved, so I'm marking this as resolved.

carlescufi replied:
I actually disagree with this, this should be `SYS_`, because we have an upcoming PR that will state that prefixes will match the subsystem they are in. In this case this is `sys/` and the functions begin with `sys_` so the macros should beging with `SYS_` as well. @andyross please take another look.

cfriedt replied:
Sure - given that there was no further input here, I was inclined to leave it, since it was approved, but I can definitely  revisit.

aescolar commented:
The test was in really bad condition and being filtered by default for the integration platform, which I guess means it did not run in CI.
Once the filter is fixed the test did not build in several configurations and when it did it failed.
I fixed the majority of the issues in this commit, but the speed optimized cases still assert (I did not dig into it):
https://github.com/aescolar/zephyr/commit/fc8e3fb44b2c14ce5267d2e131036f25116f93ee
You are welcome/I'd encourage you to take this fixes into your PR.

cfriedt commented:
> The test was in really bad condition and being filtered by default for the integration platform, which I guess means it did not run in CI.

Oh weird. So I guess maybe specifying `CONFIG_PICOLIBC=y` might have been best to keep. I could probably restrict that to a board/.conf file.

> Once the filter is fixed the test did not build in several configurations and when it did it failed.

Yeah, I can imagine that would be the case with host libc forced to 32-bit mode without the additional configuration option for 64-bit time_t in your other PR.

I think I would prefer to keep these tests limited to 64-bit `time_t` and, as a project, we should probably standardize on that.

> I fixed the majority of the issues in this commit, but the speed optimized cases still assert (I did not dig into it):

I'll take a look.

> You are welcome/I'd encourage you to take this fixes into your PR.

Thanks - I'll probably take some bits and pieces.

cfriedt commented on include/zephyr/sys/timeutil.h:
```diff
@@ -614,123 +607,60 @@ static inline bool timespec_equal(const struct timespec *a, const struct timespe
-			};
-		}
-	}
-
-#endif
+	return timespec_to_timeout_rem(ts, NULL);
```
It might even be better off to just add the `rem` parameter to `timespec_to_timeout()` and ensure it is consistent within one release.

cfriedt commented on tests/lib/timespec_util/src/main.c:
```diff
@@ -329,54 +426,140 @@ ZTEST(timeutil_api, test_timespec_from_timeout)
+					 (long long)tspec->tspec.tv_nsec);
+			}
+
 			/* no saturation / exact match */
```
```suggestion
```

andyross reviewed (dismissed):
That was my only real complaint.  This all looks reasonable.

aescolar commented:
> Yeah, I can imagine that would be the case with host libc

@cfriedt The tests fail to build also for qemu and for native_sim//64. You don't need to guess or take my word for it. You can just try them.

cfriedt commented:
Just editing (what I hope to be) my last revision on this PR...
All test cases passing on every platform of interest.

```shell
$ twister -p native_sim/native/64 -T tests/lib/timespec_util/
INFO    - Building initial testsuite list...
INFO    - Writing JSON report /home/cfriedt/zephyrproject/zephyr/twister-out/testplan.json
INFO    - JOBS: 32
INFO    - Adding tasks to the queue...
INFO    - Added initial list of jobs to queue
INFO    - Total complete:   12/  12  100%  built (not run):    0, filtered:    0, failed:    0, error:    0
INFO    - 12 test scenarios (12 configurations) selected, 0 configurations filtered (0 by static filter, 0 at runtime).
INFO    - 12 of 12 executed test configurations passed (100.00%), 0 built (not run), 0 failed, 0 errored, with no warnings in 43.61 seconds.
INFO    - 132 of 132 executed test cases passed (100.00%) on 1 out of total 1115 platforms (0.09%).
INFO    - 12 test configurations executed on platforms, 0 test configurations were only built.
INFO    - Saving reports...
INFO    - Writing JSON report /home/cfriedt/zephyrproject/zephyr/twister-out/twister.json
INFO    - Writing xunit report /home/cfriedt/zephyrproject/zephyr/twister-out/twister.xml...
INFO    - Writing xunit report /home/cfriedt/zephyrproject/zephyr/twister-out/twister_report.xml...
INFO    - Run completed
```

```shell
$ twister -p native_sim -T tests/lib/timespec_util/
INFO    - Using Ninja..
INFO    - Zephyr version: v4.2.0-rc2-42-gda53b26af51d
INFO    - Using 'zephyr' toolchain.
INFO    - Building initial testsuite list...
INFO    - Writing JSON report /home/cfriedt/zephyrproject/zephyr/twister-out/testplan.json
INFO    - JOBS: 32
INFO    - Adding tasks to the queue...
INFO    - Added initial list of jobs to queue
INFO    - Total complete:   12/  12  100%  built (not run):    0, filtered:    0, failed:    0, error:    0
INFO    - 12 test scenarios (12 configurations) selected, 0 configurations filtered (0 by static filter, 0 at runtime).
INFO    - 12 of 12 executed test configurations passed (100.00%), 0 built (not run), 0 failed, 0 errored, with no warnings in 43.96 seconds.
INFO    - 132 of 132 executed test cases passed (100.00%) on 1 out of total 1115 platforms (0.09%).
INFO    - 12 test configurations executed on platforms, 0 test configurations were only built.
INFO    - Saving reports...
INFO    - Writing JSON report /home/cfriedt/zephyrproject/zephyr/twister-out/twister.json
INFO    - Writing xunit report /home/cfriedt/zephyrproject/zephyr/twister-out/twister.xml...
INFO    - Writing xunit report /home/cfriedt/zephyrproject/zephyr/twister-out/twister_report.xml...
INFO    - Run completed
```

```shell
$ twister -p qemu_riscv64 -T tests/lib/timespec_util/
INFO    - Using Ninja..
INFO    - Zephyr version: v4.2.0-rc2-42-gda53b26af51d
INFO    - Using 'zephyr' toolchain.
INFO    - Building initial testsuite list...
INFO    - Writing JSON report /home/cfriedt/zephyrproject/zephyr/twister-out/testplan.json
INFO    - JOBS: 32
INFO    - Adding tasks to the queue...
INFO    - Added initial list of jobs to queue
INFO    - Total complete:   12/  12  100%  built (not run):    0, filtered:    0, failed:    0, error:    0
INFO    - 12 test scenarios (12 configurations) selected, 0 configurations filtered (0 by static filter, 0 at runtime).
INFO    - 12 of 12 executed test configurations passed (100.00%), 0 built (not run), 0 failed, 0 errored, with no warnings in 18.09 seconds.
INFO    - 132 of 132 executed test cases passed (100.00%) on 1 out of total 1115 platforms (0.09%).
INFO    - 12 test configurations executed on platforms, 0 test configurations were only built.
INFO    - Saving reports...
INFO    - Writing JSON report /home/cfriedt/zephyrproject/zephyr/twister-out/twister.json
INFO    - Writing xunit report /home/cfriedt/zephyrproject/zephyr/twister-out/twister.xml...
INFO    - Writing xunit report /home/cfriedt/zephyrproject/zephyr/twister-out/twister_report.xml...
INFO    - Run completed
```

danieldegrasse commented on tests/lib/timespec_util/testcase.yaml:
```diff
@@ -8,29 +8,85 @@ common:
 tests:
   libraries.timespec_utils: {}
-  libraries.timespec_utils.speed:
+  libraries.timespec_utils.timeout_32bit:
```
This is fixed on the latest push for me

aescolar reviewed (changes requested):
You have been provided all necessary changes to fix this code so it works also with other time_t sizes. Please fix it.
https://github.com/zephyrproject-rtos/zephyr/pull/92709#issuecomment-3057579685

aescolar commented:
@cfriedt in the previous to the last commit there has been both quite significant changes to the PR content and a rebase in 1.
Doing the rebase and a change in one push makes it quite much more difficult for reviewers to handle it.
It is much nicer to do two separate pushes one with a rebase and the other with the changes in the code.

cfriedt commented:
* added modulo operator back, although it was completely unnecessary
* corrected code formatting

cfriedt commented:
@andyross - would be great if you could re-approve.

@danieldegrasse @dkalowsk - it would be really quite good to get this `Priority: high` fix in so that `v4.2.0` is not released with bugs for all platforms 🤷‍♂️ 

I've made quite a few of the recommended changes by Alberto. Coverage and bisectability tests show that it's safe.

I think it's ultimately up to you as release managers.

aescolar commented:
@cfriedt

The changes to support any time_t size are trivial, and you got them in https://github.com/zephyrproject-rtos/zephyr/pull/92709#issuecomment-3057579685

How I see it, you have two options:

a) you apply those changes, this code supports and time_t size, this PR can be merged and we move on.
b) you continue arguing. But then you can expect me to continue requesting changes due to the issues already mentioned.

aescolar commented on include/zephyr/sys/clock.h:
```diff
@@ -166,6 +166,49 @@ typedef struct {
 /* added tick needed to account for tick in progress */
 #define _TICK_ALIGN 1
 
+/* Maximum and minimum value TIME_T can hold (accounting for 32-bit time_t in glibc.h) */
```
```suggestion
/* Maximum and minimum value TIME_T can hold */
```
This support any time_t not just 32bits

aescolar replied:
Still applies

cfriedt commented on include/zephyr/sys/clock.h:
```diff
@@ -166,6 +166,49 @@ typedef struct {
+#define K_TICK_MAX ((k_ticks_t)(IS_ENABLED(CONFIG_TIMEOUT_64BIT) ? INT64_MAX : UINT32_MAX - 1))
+
+/* The semantic equivalent of K_NO_WAIT but expressed as a timespec object*/
+#define K_TS_NO_WAIT K_TICKS_TO_TIMESPEC(0)
```
Sure - I can do that.

aescolar commented:
> If this is not merged within this release cycle, then we will need to rename timespec_to_timeout(req, rem) to something like timespec_to_timeout_remainder(req, rem). 

We can just backport this change to 4.2.
Our deprecation policy is for API we can expect users to have started using, we do not need to follow it for something which has been a couple of weeks in the tree.

About this code today requiring a 64bit time_t, it seems you are misunderstanding my motivation to ask for any time_t to be suported.
So I'll try to elaborate:

The size of time_t is decided by the C library given the platform.
We support proprietary toolchains (which tend to have their own C libraries), we also support user provided C libraries.
In the tree we default to Picolibc for most platforms, but that does not mean users can only use pico or that they do.
We also support out of tree architectures.

Several of our users and contributors (including TSC members) have their own out of tree architectures, toolchains and C libraries, and/or have clients who do.

For processors focused on audio processing 24/48bit architectures are very typical.
Where native integer types are 24/48 bits, you can also expect the c library to define a time_t of 48bits which does not have any Y2038 issue.

Many users will not want to care at all about time_t/timespec either because they do not have a need for that kind type, or because they are all together avoiding it due to the trouble around time_t and the functions that use it (see your own MISRA reference as an example).
Due to this, and the presence of legacy libraries which either assume time_t is a given size or are prebuilt for a given size, some libCs will keep using an old time_t size by default.
And many of those libC users just do not care or need to care.

This code (lib/os/clock.c and timeutil) is built always, and so are these new (for a couple of weeks) functions. Even for users do not have any interest in using time_t or timespec. These users will probably be a bit confused about what part of the zephyr code they may use could be using these or not.
A test failing in their platform with a requirement for time_t being a given size will confuse them.

Quite many users are using the native_simulator based targets for host testing, and prefer to build with their host libc (glibc typically) so they can interact with the host easier. A lot of this "host interaction" happens by using other host prebuilt libraries.
Forcing glibc to build with a 64bit time_t breaks the ABI for calls using that type or derived types, and that is a nasty bug to debug. If the library they use is prebuilt, the user may not have an option to rebuild it with a 64bit time_t.

In general many of Zephyr developers just want to be able to test their code with any libC and don't want unnecessary constraints set on them.
They don't want to depend on code that only works with a given size of time_t or only works with some libCs if they happen to be using it as part of their testing on their target or host.

You seem to think that the Zephyr project can easily force time_t to be 64bits. But it can't, because its size is defined by the libC. All the Zephyr project can do is make it more difficult to work with other time_t sizes and take time and frustrate our users/developers.

What I'm requesting is that this code would support any time_t size. Be it 32, 48, 64 or whatever the libC and user decides.
You have already done most of the changes for this to be the case (most replacements of UINT64_MAX/MIN with SYS_TIME_T_MAX/MIN).
The remaining changes are trivial. And none of them have any drawback for us or our users. Pico will still default to the 64bit time_t. For other libCs the users can still chose the size they want. If they care about time_t and the year 2038 issue they just need to ensure it is more than 32bits, which is the default in the tree for embedded targets anyhow.

If you do this change, the code supports any libc, so you can remove the filter on the library on this code, and stop selecting pico (if you want). And any other developer whose code uses this code can also avoid having any constraint about the libC choice or time_t size in their testing.

I just don't see any drawback in supporting any time_t size, but quite a lot of drawbacks of not supporting it.
While supporting any time_t size does not mean we force or default to using any time_t size in embedded products that has any issue.

cfriedt commented:
2025-07-15: release meeting
 - relatively new api, not stable, does not need to follow the deprecation process
 - can be merged after the release
 - backport to `v4.2.1`

cfriedt commented:
@aescolar - do you mind unblocking this PR? I don't think there is any reason for it to be blocked post-release.

If you would like to create a follow-up PR for the additional features you want, feel free.

@nashif, @andyross, @peter-mitsis - if any of you are willing to (re-)approve, I would appreciate it.

I'll rebase this momentarily to refresh it, in any case.

cfriedt commented:
* rebased to avoid unrelated twister failure

cfriedt commented:
This PR is just about approved, but illustrates that the 32-bit `time_t` issue is mostly external to this PR which is mainly focused on fixing bugs.
https://github.com/zephyrproject-rtos/zephyr/pull/93535

Hopefully today, this PR can be addressed in the arch WG.

cfriedt commented:
> @cfriedt there is several issues which have been pointed in this PR reviews which have not been addressed. I reminded you 2 days ago about one which is evident as the first thing in the diff: [#92709 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/92709#discussion_r2238841195) Addressing those issues is probably a faster way to get this merged that ignoring them and asking others to do things.

Sure - I can update the comment.

cfriedt commented:
Arch Review 2025-07-31:
* seems reasonable to keep the scope of this PR narrow
* 32-bit `time_t` can be added to the testsuite after #92709 is merged
* @cfriedt to modify comment as requested by @aescolar

aescolar commented:
@cfriedt
> Arch Review 2025-07-31:

What you did here is not ok:
* You cannot take over an unrelated cancelled meeting instance, self appoint yourself as chair and discuss whatever other topic.
* Bringing your own PR for discussion to a meeting without an agenda or involving the relevant parties (the maintainers and reviewers) is not helpful.
  * During that one sided presentation, claiming half truths or downright falsehoods as true, stating what reviewers want or intend when not true, and stating that others have done what they have not done is not constructive.
* After all this, you cannot post a [comment](https://github.com/zephyrproject-rtos/zephyr/pull/92709#issuecomment-3140403762) like if the matter had been properly discussed by the necessary parties and a resolution was reached in the Arch WG forum.

All this is just subverting the processes we have, creating unnecessary confusion, and wasting time.
Please refrain from doing this.

cfriedt commented:
> @cfriedt
> 
> > Arch Review 2025-07-31:
> 
> What you did here is not ok:
>
> * You cannot take over an unrelated cancelled meeting instance, self appoint yourself as chair and discuss whatever other topic.

What the actual F? WHAT YOU DID here is not ok.

What YOU did here, in this comment, is ridiculous, accusatory, defamatory, and inflammatory!

The topic was brought up in Tuesday's arch meeting. The Thursday arch meeting was not cancelled. I discussed with @carlescufi in advance. He was unable to published the meeting agenda in advance due to a family emergency. There was no conspiracy as you seem to believe.

At no time did I make any claim that I was appointing myself as chair and I said I would be happy to leave if nobody showed up within 5 minutes.

The meeting proceeded as usual, was attended by @bjarki-andreasen, @danieldegrasse, @dleach02, @MaureenHelm, and a couple of others.

> * Bringing your own PR for discussion to a meeting without an agenda or involving the relevant parties (the maintainers and reviewers) is not helpful.

In fact it was quite helpful, and I followed the suggested provided by the attendees.

>   * During that one sided presentation, claiming half truths or downright falsehoods as true, stating what reviewers want or intend when not true, and stating that others have done what they have not done is not constructive.

What was a downright falsehood??

> * After all this, you cannot post a [comment](https://github.com/zephyrproject-rtos/zephyr/pull/92709#issuecomment-3140403762) like if the matter had been properly discussed by the necessary parties and a resolution was reached in the Arch WG forum.

I took notes of what was discussed in the meeting.

> All this is just subverting the processes we have, creating unnecessary confusion, and wasting time. Please refrain from doing this.

I have not, nor is anyone, attempting to subvert any process.

These constant personal attacks are getting really old Alberto.

STOP!!!

nashif commented:
A few general comments

> As the deadline for v4.2.0 had approached, and since it had been previously approved, I dismissed Alberto's comment as there was no further maintainer feedback on the subject.

You should not dismiss reviews and requests for changes from a reviewer because someone else previously approved. Approval becomes void once the content PR was changed. Even with many existing approvals, a new review or request for changes should be dealt with and addressed. Who approved and the number for approval do not make a request for change or additional reviews irrelevant. 

> The topic was brought up in Tuesday's arch meeting. The Thursday arch meeting was not cancelled. I discussed with @carlescufi in advance. He was unable to published the meeting agenda in advance due to a family emergency. There was no conspiracy as you seem to believe.

why was this topic brought in the arch WG? We do not discuss PRs in the WG unless they are being escalated. As far as I can tell, the PR was still going and the review process itself was not exhausted to justify escalation. But even if escalation was needed, IMO it was abrupt and rash and did not follow the escalation process, i.e. the least I would have expected for all parties involved to be invited and present in the discussion. Were @aescolar or any of the assignees of this PR notified or made aware of the escalation or discussion?  Maybe the hope was they will join anyways, but I would not have gone through without everyone present, from the docs:

```
Before escalation of technical disagreements, follow the steps below:

- Resolve in the PR among assignee, maintainers and reviewer.

  - Assignee to act as moderator if applicable.

- Optionally resolve in the next `Zephyr Dev Meeting`_  meeting with more
  Maintainers and project stakeholders.

  - The involved parties and the Assignee to be present when the  issue is
    discussed.
```

> What you did here is not ok:
> 
> * You cannot take over an unrelated cancelled meeting instance, self appoint yourself as chair and discuss whatever other topic.
> * Bringing your own PR for discussion to a meeting without an agenda or involving the relevant parties (the maintainers and reviewers) is not helpful.
>   
>   * During that one sided presentation, claiming half truths or downright falsehoods as true, stating what reviewers want or intend when not true, and stating that others have done what they have not done is not constructive.
> * After all this, you cannot post a [comment](https://github.com/zephyrproject-rtos/zephyr/pull/92709#issuecomment-3140403762) like if the matter had been properly discussed by the necessary parties and a resolution was reached in the Arch WG forum.
> 
> All this is just subverting the processes we have, creating unnecessary confusion, and wasting time. Please refrain from doing this.

@aescolar Given what I said above, this should have been brought up with WG chair (@carlescufi ) and dealt with outside of this PR, we should avoid this from happening again and we need to overhaul out escalation process, which seems a bit outdated. 
That discussion in the WG given our processes mentioned before does not carry any weight absent all stakeholders. 


@cfriedt 
> They say that the only way to win at a game that is not fair is to simply not play. That is why I have closed this PR.

This is unfortunate.

ycsin commented:
It is a regret to see how this bug-fixing PR has ended up, is there a way that we can move this forward productively? From what I can tell based on the discussions above (please correct me if I'm wrong) there are only 2 outstanding concerns:

1. Use of `SYS_*` instead of `K_*` for the macros to align with current direction
2. Whether `tests/lib/timespec_util` should be modified to support non 64-bit `time_t`

My 2 cents:
1. Let's update them in this PR if most of the reviewers think that it is better than changing them altogether in a later treewide PR
2. The testcase was made for 64-bit all this while, it should be reasonable to keep it that way in this PR. That said, another PR can be submitted to support non 64-bit `time_t` for the testcase

If @cfriedt, @aescolar and other stakeholders don't mind, I can create another PR to help get this bug-fix through!
